### PR TITLE
103926 Message subscriber infrastructure/Project Subscription

### PR DESCRIPTION
**Pull request by TordJoranger**

### CODE REVIEW

File: src/Equinor.ProCoSys.Completion.Infrastructure/CompletionContext.cs [hunk 1]
```diff
@@ -79,13 +79,9 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
...
 
-    private static void ConfigureMassTransitOutBox(ModelBuilder modelBuilder)
-    {
-        modelBuilder.AddInboxStateEntity();
-        modelBuilder.AddOutboxMessageEntity();
-        modelBuilder.AddOutboxStateEntity();
-    }
-
+    private static void ConfigureOutBoxPattern(ModelBuilder modelBuilder) 
+        => modelBuilder.AddTransactionalOutboxEntities();
```
Comment (TordJoranger) [inline src/Equinor.ProCoSys.Completion.Infrastructure/CompletionContext.cs:83]: this does the same as the 3 lines over, convince method from MT

File: src/Equinor.ProCoSys.Completion.Domain/AggregateModels/ProjectAggregate/Project.cs [hunk 2]
```diff
@@ -20,18 +20,20 @@ protected Project()
...
     }
 
-    public Project(string plant, Guid guid, string name, string description)
+    public Project(string plant, Guid guid, string name, string description, DateTime lastUpdated)
         : base(plant)
     {
         Guid = guid;
         Name = name;
         Description = description;
+        Pcs4LastUpdated = lastUpdated;
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Domain/AggregateModels/ProjectAggregate/Project.cs:29]: ProCoSys4LastUpdated

File: src/Equinor.ProCoSys.Completion.Domain/AggregateModels/ProjectAggregate/Project.cs [hunk 3]
```diff
@@ -20,18 +20,20 @@ protected Project()
...
     }
 
     // private setters needed for Entity Framework
     public string Name { get; set; }
     public string Description { get; set; }
     public bool IsClosed { get; set; }
+    public DateTime Pcs4LastUpdated { get; set; }
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Domain/AggregateModels/ProjectAggregate/Project.cs:36]: Add .HasConversion(CompletionContext.DateTimeKindConverter) in ProjectConfiguration (see PunchItemConfiguration)

File: src/Equinor.ProCoSys.Completion.WebApi/DiModules/ApplicationModule.cs [hunk 4]
```diff
@@ -64,15 +65,37 @@ public static void AddApplicationModules(this IServiceCollection services, IConf
...
+                    return opts;
+                });
+                cfg.SubscriptionEndpoint("completion_project","project", e =>
+                {
+                    e.ClearSerialization();
+                    e.UseRawJsonSerializer();
+                    e.UseRawJsonDeserializer();
+                    e.ConfigureConsumer<ProjectEventConsumer>(context);
+                    e.ConfigureConsumeTopology = false;
+                    e.PublishFaults = false; //I didn't get this to work, I think it tried to publish to endpoint that already exists in different context or something, we're logging errors anyway.
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.WebApi/DiModules/ApplicationModule.cs:97]: remove blank line

File: src/Equinor.ProCoSys.Completion.WebApi/DiModules/ApplicationModule.cs [hunk 5]
```diff
@@ -64,15 +65,37 @@ public static void AddApplicationModules(this IServiceCollection services, IConf
...
+            x.UsingAzureServiceBus((context,cfg) =>
             {
                 var connectionString = configuration.GetConnectionString("ServiceBus");
                 cfg.Host(connectionString);
                 
                 cfg.MessageTopology.SetEntityNameFormatter(new ProCoSysKebabCaseEntityNameFormatter());
                 
+                cfg.ConfigureJsonSerializerOptions(opts =>
+                {
+                    opts.Converters.Add(new OracleGuidConverter());
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.WebApi/DiModules/ApplicationModule.cs:86]: why do we need the OracleGuidConverter?

File: src/Equinor.ProCoSys.Completion.Domain/AggregateModels/ProjectAggregate/Project.cs (hunk 2 above)
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Domain/AggregateModels/ProjectAggregate/Project.cs:29]: Check for Kind.Utc, throw exception if not ... search through solution and replace , DateTime.Now with , DateTime.UtcNow where used in Project constructor (Seeder.cs ++)

File: src/Equinor.ProCoSys.Completion.WebApi/Synchronization/ProjectEventConsumer.cs [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    private readonly ILogger<ProjectEventConsumer> _logger; 
+    private readonly IProjectRepository _projectRepository;
+    private readonly IUnitOfWork _unitOfWork;
+    private readonly ICurrentUserSetter _currentUserSetter;
+    private readonly IOptions<CompletionAuthenticatorOptions> _options;
+    
+    public ProjectEventConsumer(ILogger<ProjectEventConsumer> logger, 
+        IProjectRepository projectRepository, 
+        IUnitOfWork unitOfWork, 
+        ICurrentUserSetter currentUserSetter, IOptions<CompletionAuthenticatorOptions> options)
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.WebApi/Synchronization/ProjectEventConsumer.cs:25]: split into 2 lines

File: src/Equinor.ProCoSys.Completion.WebApi/Synchronization/ProjectEventConsumer.cs [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    {
+        _logger = logger;
+        _projectRepository = projectRepository;
+        _unitOfWork = unitOfWork;
+        _currentUserSetter = currentUserSetter;
+        _options = options;
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.WebApi/Synchronization/ProjectEventConsumer.cs:31]: _completionApiObjectId = _options.Value.CompletionApiObjectId

File: src/Equinor.ProCoSys.Completion.WebApi/Synchronization/ProjectEventConsumer.cs (hunk 6 above)
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.WebApi/Synchronization/ProjectEventConsumer.cs:25]: Use IOptionsMonitor<>

File: src/Equinor.ProCoSys.Completion.WebApi/Synchronization/ProjectEventConsumer.cs [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+            var project = await _projectRepository.GetAsync(projectEvent.ProCoSysGuid, context.CancellationToken);
+            _projectRepository.Remove(project);
+        }
+        
+        else if(await _projectRepository.ExistsAsync(projectEvent.ProCoSysGuid, context.CancellationToken))
+        {
+            var project = await _projectRepository.GetAsync(projectEvent.ProCoSysGuid, context.CancellationToken);
+            if (project.Pcs4LastUpdated > projectEvent.LastUpdated)
+            {
+                _logger.LogWarning("Project Message Ignored because a newer LastUpdated already exits in db:" +
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.WebApi/Synchronization/ProjectEventConsumer.cs:54]: \n behind db: and remove space in front of MessageId at next line
Split " MessageId: {MessageId} \n ProCoSysGuid {ProCoSysGuid} \n " into 2 lines, with \n as last char

File: src/Equinor.ProCoSys.Completion.WebApi/Synchronization/ProjectEventConsumer.cs [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+            context.MessageId, projectEvent.ProjectName);
+    }
+
+    private static void MapFromEventToProject(IProjectEventV1 projectEvent, Project project)
+    {
+        project.IsClosed = projectEvent.IsClosed;
+        project.Name = projectEvent.ProjectName;
+        project.Pcs4LastUpdated = projectEvent.LastUpdated;
+        
+        if (projectEvent.Description != null)
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.WebApi/Synchronization/ProjectEventConsumer.cs:81]: is not null

File: src/Equinor.ProCoSys.Completion.WebApi/Synchronization/ProjectEventConsumer.cs [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        {
+            project.Description = projectEvent.Description;
+        }
+    }
+
+    //12/12/2023 We use name as description if description is missing. Its a super edge case and pcs4 does not have any projects with description is null today
+     private static Project CreateProjectEntity(IProjectEventV1 projectEvent) 
+         => new(projectEvent.Plant, projectEvent.ProCoSysGuid, projectEvent.ProjectName,projectEvent.Description?? projectEvent.ProjectName, projectEvent.LastUpdated);
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.WebApi/Synchronization/ProjectEventConsumer.cs:89]: long line, split so one param on each line

File: src/Equinor.ProCoSys.Completion.WebApi/Synchronization/ProjectEventConsumer.cs [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+public record ProjectEvent(string EventType, string? Description,
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.WebApi/Synchronization/ProjectEventConsumer.cs:92]: move to its own file

File: src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/ProjectEventConsumerTests.cs [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+namespace Equinor.ProCoSys.Completion.WebApi.Tests.Synchronization;
+
+[TestClass]
+public class ProjectEventConsumerTests
+{
+    private readonly IProjectRepository _projectRepoMock = Substitute.For<IProjectRepository>();
+    private readonly IUnitOfWork _unitOfWorkMock = Substitute.For<IUnitOfWork>();
+    private readonly ProjectEventConsumer _projectEventConsumer;
+    private readonly IOptions<CompletionAuthenticatorOptions> _optionsMock = Substitute.For<IOptions<CompletionAuthenticatorOptions>>();
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/ProjectEventConsumerTests.cs:22]: long line

File: src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/ProjectEventConsumerTests.cs [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+    private readonly ConsumeContext<ProjectEvent> _contextMock = Substitute.For<ConsumeContext<ProjectEvent>>();
+
+    public ProjectEventConsumerTests() =>
+        _projectEventConsumer = new ProjectEventConsumer(Substitute.For<ILogger<ProjectEventConsumer>>(), _projectRepoMock,
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/ProjectEventConsumerTests.cs:26]: one param on each line

File: src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/ProjectEventConsumerTests.cs [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+[TestClass]
+public class ProjectEventConsumerTests
+{
+    private readonly IProjectRepository _projectRepoMock = Substitute.For<IProjectRepository>();
+    private readonly IUnitOfWork _unitOfWorkMock = Substitute.For<IUnitOfWork>();
+    private readonly ProjectEventConsumer _projectEventConsumer;
+    private readonly IOptions<CompletionAuthenticatorOptions> _optionsMock = Substitute.For<IOptions<CompletionAuthenticatorOptions>>();
+    private readonly ConsumeContext<ProjectEvent> _contextMock = Substitute.For<ConsumeContext<ProjectEvent>>();
+
+    public ProjectEventConsumerTests() =>
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/ProjectEventConsumerTests.cs:25]: who both constructor AND Setup?
Prefer to remove constructor and move init of _projectEventConsumer  to Setup

File: src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/ProjectEventConsumerTests.cs [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+                "ProjectName", null);
+        var contextMock = Substitute.For<ConsumeContext<ProjectEvent>>();
+        contextMock.Message.Returns(testEvent);
+        
+        //Act and Assert
+        await Assert.ThrowsExceptionAsync<Exception>(() => _projectEventConsumer.Consume(contextMock),"Message is missing ProCoSysGuid");
+    }
+    
+    [TestMethod]
+    public async Task Consume_ShouldRemoveProject_IfMessageContainsDelete()
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/ProjectEventConsumerTests.cs:50]: IfBehaviorIsDelete

File: src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/ProjectEventConsumerTests.cs [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+        _optionsMock.Value.Returns(new CompletionAuthenticatorOptions
+        {
+            CompletionApiObjectId = new Guid()
+        });
+
+    [TestMethod]
+    public async Task Consume_ShouldThrowException_IfNoProCoSysGuid()
+    {
+        //Arrange
+        var testEvent = new ProjectEvent("", "project", false, DateTime.MinValue, "plant", Guid.Empty,
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/ProjectEventConsumerTests.cs:40]: Prefer to see one param on each line (when not all params fit on one line) ... this through whole this class

File: src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/ProjectEventConsumerTests.cs [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+        //Act
+        await _projectEventConsumer.Consume(_contextMock);
+        
+        //Assert
+        await _projectRepoMock.Received(1).GetAsync(proCoSysGuidForProjectToUpdate, default);
+        _projectRepoMock.Received(0).Remove(project);
+        _projectRepoMock.Received(0).Add(project);
+        await _unitOfWorkMock.Received(0).SaveChangesAsync(default);
+    }
+
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/ProjectEventConsumerTests.cs:91]: I like to see more positive tests:
1) Test the update
2) Test the add with description
3) Test the add without description (the edge case)

**Review by janid1967 — CHANGES_REQUESTED** [review]
See comments

File: src/Equinor.ProCoSys.Completion.WebApi/Synchronization/ProjectEventConsumer.cs (hunk 10 above)
Comment (TordJoranger) [reply src/Equinor.ProCoSys.Completion.WebApi/Synchronization/ProjectEventConsumer.cs:89]: la gå :)

File: src/Equinor.ProCoSys.Completion.Domain/AggregateModels/ProjectAggregate/Project.cs (hunk 2 above)
Comment (TordJoranger) [reply src/Equinor.ProCoSys.Completion.Domain/AggregateModels/ProjectAggregate/Project.cs:29]: we discussed this, wont fix for now, but will add explaining comment

File: src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/ProjectEventConsumerTests.cs (hunk 14 above)
Comment (TordJoranger) [reply src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/ProjectEventConsumerTests.cs:26]: because of nullability constraints

File: src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/ProjectEventConsumerTests.cs [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+        _projectRepoMock
+            .When(x => x.Add(Arg.Any<Project>()))
+            .Do(callInfo =>
+            {
+                _projectAddedToRepository = callInfo.Arg<Project>();
+            });
+    }
+
+    [TestMethod]
+    public async Task Consume_WorksAsExpected_WhenAddingNewProject()
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/ProjectEventConsumerTests.cs:44]: Consume_ShouldAddNewProject_WhenProjectNotExists

File: src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/ProjectEventConsumerTests.cs [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+        Assert.AreEqual(guid,_projectAddedToRepository.Guid);
+        Assert.AreEqual(false,_projectAddedToRepository.IsClosed);
+        Assert.AreEqual(lastUpdated,_projectAddedToRepository.ProCoSys4LastUpdated);
+        Assert.AreEqual(Description,_projectAddedToRepository.Description);
+        Assert.AreEqual(ProjectName,_projectAddedToRepository.Name);
+        await _unitOfWorkMock.Received(1).SaveChangesAsync(default);
+    }
+    
+    [TestMethod]
+    public async Task Consume_WorksAsExpected_WhenUpdatingProject()
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/ProjectEventConsumerTests.cs:72]: Consume_ShouldUpdateExistingProject_WhenProjectExists

File: src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/ProjectEventConsumerTests.cs [hunk 20]
```diff
@@ -0,0 +1,187 @@
...
+        Assert.AreEqual(guid,projectToUpdate.Guid);
+        Assert.AreEqual(false,projectToUpdate.IsClosed);
+        Assert.AreEqual(lastUpdated,projectToUpdate.ProCoSys4LastUpdated);
+        Assert.AreEqual(Description,projectToUpdate.Description);
+        Assert.AreEqual(ProjectName,projectToUpdate.Name);
+        await _unitOfWorkMock.Received(1).SaveChangesAsync(default);
+    }
+    
+    [TestMethod]
+    public async Task Consume_UsedProjectNameAsDescription_WhenDescriptionIsNull()
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/ProjectEventConsumerTests.cs:101]: Consume_ShouldUseProjectNameAsDescription_WhenDescriptionIsNull()

**Review by janid1967 — APPROVED** [review]
Approved with minor rename suggestions of test methods
